src/plugins/advanced_settings/public/types.ts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { componentRegistry } from './component_registry';
+
+export interface AdvancedSettingsSetup {
+  componentRegistry: typeof componentRegistry;
+}
+export interface AdvancedSettingsStart {
+  componentRegistry: typeof componentRegistry;
+}
```

Comment:
We probably don't want these interfaces to be the same since it is a registry, right?

i.e. on Setup maybe only `register` should be available, and on Start only `get`, so that nobody is trying to register after setup

Reply:
It looks like its sufficient to provide `register` access in the setup lifecycle although it doesn't seem  necessary to me - but I might be missing something. That said, I'm a little uncertain how this relates to legacy code - I'd be using `register` after setup is complete, correct?

Reply:
The reasoning behind doing it this way in other registries has been that, if you only allow registering in the `setup` lifecycle, by the time you get to `start` you can guaranteed that all items that are going to be registered are already present, and can then safely use `get` to retrieve anything from the registry.

This avoids the scenario where you may try to do something with the items in a registry which has not been fully populated yet.

It's a little weird in this case because I believe the only place the registry is being read from is in the advanced settings UI itself, which means new registry items will likely just re-render in React if they come in late. So I'm not sure if it would have a material impact in this particular plugin or not; I'm mostly bringing it up for the sake of consistency with how we've dealt with registries in other plugins. Examples of places where we've done this are expressions, inspector, management.

Reply:
Hm, what about license changes? How does that play with the setup / start, register / get pattern?